ext/hook/uhook.c

```diff
@@ -0,0 +1,331 @@
+
+    HashTable *ht = emalloc(sizeof(*ht));
+    zend_hash_init(ht, num_args, NULL, ZVAL_PTR_DTOR, 0);
+
+    if (!num_args) {
+        return ht;
+    }
+
+    zval *p = EX_VAR_NUM(0);
+    zend_function *func = EX(func);
+    ht->nNumOfElements = num_args;
+
+    zend_hash_real_init(ht, 1);
+    ZEND_HASH_FILL_PACKED(ht) {
+        if (EX(func)->type == ZEND_USER_FUNCTION) {
+            uint32_t first_extra_arg = MIN(num_args, func->op_array.num_args);
+
+            for (zval *end = p + first_extra_arg; p < end; ++p) {
+                if (Z_OPT_REFCOUNTED_P(p)) {
+                    Z_ADDREF_P(p);
+                }
+                ZEND_HASH_FILL_ADD(p);
+            }
+
+            p = EX_VAR_NUM(func->op_array.last_var + func->op_array.T);
+            num_args -= first_extra_arg;
+        }
+
+        for (zval *end = p + num_args; p < end; ++p) {
+            if (Z_OPT_REFCOUNTED_P(p)) {
+                Z_ADDREF_P(p);
+            }
+            ZEND_HASH_FILL_ADD(p);
+        }
+    }
+    ZEND_HASH_FILL_END();
+
+    return ht;
+}
+
+static void dd_uhook_call_hook(zend_execute_data *execute_data, zend_object *closure, zend_object *hook_data) {
+    zval closure_zv, hook_data_zv;
+    ZVAL_OBJ(&closure_zv, closure);
+    ZVAL_OBJ(&hook_data_zv, hook_data);
+
+    bool has_this = getThis() != NULL;
+    zval rv;
+    zai_symbol_call(has_this ? ZAI_SYMBOL_SCOPE_OBJECT : ZAI_SYMBOL_SCOPE_GLOBAL, has_this ? &EX(This) : NULL, ZAI_SYMBOL_FUNCTION_CLOSURE, &closure_zv, &rv, 1, &hook_data_zv);
+    zval_ptr_dtor(&rv);
+}
+
+static bool dd_uhook_begin(zend_ulong invocation, zend_execute_data *execute_data, void *auxiliary, void *dynamic) {
+    (void) invocation;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    dyn->hook_data = zend_objects_new(ddtrace_hook_data_ce);
+    object_properties_init(dyn->hook_data, ddtrace_hook_data_ce);
+
+    ZVAL_LONG(ddtrace_hookdata_property_id(dyn->hook_data), def->id);
+    ZVAL_ARR(ddtrace_hookdata_property_args(dyn->hook_data), dd_uhook_collect_args(execute_data));
+
+    if (def->begin && !def->running) {
+        def->running = true;
+        dd_uhook_call_hook(execute_data, def->begin, dyn->hook_data);
+        def->running = false;
+    }
+
+    return true;
+}
+
+static void dd_uhook_end(zend_ulong invocation, zend_execute_data *execute_data, zval *retval, void *auxiliary, void *dynamic) {
+    (void) invocation;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    if (def->end && !def->running) {
+        zval tmp;
+
+        /* If the profiler doesn't handle a potential pending interrupt before
+         * the observer's end function, then the callback will be at the top of
+         * the stack even though it's not responsible.
+         * This is why the profiler's interrupt function is called here, to
+         * give the profiler an opportunity to take a sample before calling the
+         * tracing funcation.
```

Comment:
Ha, my own typo copy/pasted 🙃 , but maybe `hook` would be more precise here?

```suggestion
         * tracing hook.
```